tool/planet/reconciler.go

```diff
@@ -0,0 +1,351 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gravitational/planet/lib/constants"
+	"github.com/gravitational/planet/lib/reconcile"
+
+	"github.com/gravitational/coordinate/leader"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// Name returns the name of this plan.
+// Implements reconciler.Plan
+func (r *agentPlan) Name() string {
+	return "agent state reconciler"
+}
+
+// Creates a new plan for agent operation
+// Implements reconciler.Plan
+func (r *agentPlan) Create(ctx context.Context) (steps []reconcile.Step, err error) {
+	logger := logrus.WithField(trace.Component, r.Name())
+	if !r.electionPaused {
+		logger.Debug("Start election participation.")
+		r.addVoter(ctx)
+	} else {
+		logger.Debug("Stop election participation.")
+		r.removeVoter(ctx)
+	}
+	if r.leaderAddr == "" {
+		return steps, nil
+	}
+	if err := verifyLeaderAddr(r.leaderAddr); err != nil {
+		logger.WithError(err).Warn("Failed to check local leader DNS configuration, will update configuration.")
+		steps = append(steps, updateDNS{leaderAddr: r.leaderAddr})
+	}
+	state := activeStateActive
+	units := serviceUnits
+	if r.localAddr != r.leaderAddr {
```

Comment:
So on the leader node, the reconciler will make sure that all services (incl. docker, kubelet, etc.) are active. But on a regular node, it only makes sure that control plane components are inactive. Is this intended? Do we need to include non control plane components at all in the reconciler?

Reply:
I've revised the part as it was wrong in deciding which services to stop/start.